client/blocks/qr-code-login/index.jsx

```diff
@@ -0,0 +1,185 @@
+		return false;
+	}
+	return tokenData.expires > Date.now() / 1000;
+};
+
+const getLoginActionResponse = async ( action, args ) => {
+	const url = new URL( 'https://wordpress.com/wp-login.php' );
+	url.searchParams.append( 'action', action );
+
+	Object.keys( args ).forEach( ( key ) => {
+		url.searchParams.append( key, args[ key ] );
+	} );
+
+	const response = await fetch( url.href );
+	return await response.json();
+};
+
+const fetchQRCodeData = async ( setTokenData, tokenData, anonymousUserId ) => {
+	if ( isStillValidToken( tokenData ) ) {
+		return tokenData;
+	}
+	// tokenData is set to null initially.
+	// Lets wait till it is set to false when the local data is the just yet.
+	if ( tokenData === null ) {
+		return null;
+	}
+
+	const responseData = await getLoginActionResponse( 'qr-code-token-request-endpoint', {
+		anon_id: anonymousUserId,
+	} );
+
+	setTokenData( responseData.data );
+	setStoredItem( LOCALE_STORAGE_KEY, responseData.data );
+};
+
+const fetchAuthState = async ( setAuthState, setTokenData, tokenData, anonymousUserId ) => {
+	if ( ! tokenData ) {
+		return;
+	}
+
+	if ( ! isStillValidToken( tokenData ) ) {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+		return;
+	}
+
+	const { token, encrypted } = tokenData;
+
+	const responseData = await getLoginActionResponse( 'qr-code-authentication-endpoint', {
+		anon_id: anonymousUserId,
+		token: token,
+		data: encrypted,
+	} );
+
+	setAuthState( responseData.data );
+};
+
+function TokenQRCode( { tokenData } ) {
+	if ( ! tokenData ) {
+		return <QRCodePlaceholder />;
+	}
+	const { token, encrypted } = tokenData;
+	const imageSettings = {
+		src: qrCenter,
+		height: 64,
+		width: 64,
+		excavate: false,
+	};
+	return (
+		<QRCode
+			value={ `https://apps.wordpress.com/get/?campaign=login-qr-code#qr-code-login?token=${ token }&data=${ encrypted }` }
+			size={ 300 }
+			imageSettings={ imageSettings }
+		/>
+	);
+}
+
+function QRCodePlaceholder() {
+	return (
+		<div className="qr-code-login__placeholder">
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+			<span className="qr-code-login__corner-box"></span>
+		</div>
+	);
+}
+
+function QRCodeLogin() {
+	const translate = useTranslate();
+	const [ tokenData, setTokenData ] = useState( null );
+	const [ authState, setAuthState ] = useState( false );
+
+	const anonymousUserId = getTracksAnonymousUserId();
+
+	// Fetch QR code data.
+	useEffect( () => {
+		fetchQRCodeData( setTokenData, tokenData, anonymousUserId );
+	}, [ setTokenData, tokenData, anonymousUserId ] );
+
+	// Fetch the Auth Data.
+	useInterval( () => {
+		fetchAuthState( setAuthState, setTokenData, tokenData, anonymousUserId );
+	}, AUTH_PULL_INTERVAL );
+
+	useEffect( () => {
+		if ( authState?.auth_url ) {
+			window.location.replace( authState.auth_url );
+		}
+	}, [ authState ] );
+
+	useEffect( () => {
+		getStoredItem( LOCALE_STORAGE_KEY ).then( ( localTokenData ) =>
```

Comment:
What's the goal of this? Didn't we just do it in another effect on mount and update with `fetchQRCodeData` above?

Reply:
The goal of this is that if a user refreshed the page we do not do another request and get a new QR code. Since the user will might have scanned the code but not approved it already. This will prevent this edge case. Storing the code and using it until it was expired will prevent this case.